node/core/pvf/src/host.rs

```diff
@@ -0,0 +1,1049 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Validation host - is the primary interface for this crate. It allows the clients to enqueue
+//! jobs for PVF execution or preparation.
+//!
+//! The validation host is represented by a future/task that runs an event-loop and by a handle,
+//! [`ValidationHost`], that allows communication with that event-loop.
+
+use crate::{
+	Priority, Pvf, ValidationError,
+	artifacts::{Artifacts, ArtifactState, ArtifactId},
+	execute, prepare,
+};
+use std::{
+	collections::HashMap,
+	time::{Duration, SystemTime},
+};
+use always_assert::never;
+use async_std::{
+	path::{Path, PathBuf},
+	sync::Mutex,
+};
+use polkadot_parachain::primitives::ValidationResult;
+use futures::{
+	Future, FutureExt, SinkExt, StreamExt,
+	channel::{mpsc, oneshot},
+};
+
+/// An alias to not spell the type for the oneshot sender for the PVF execution result.
+pub(crate) type ResultSender = oneshot::Sender<Result<ValidationResult, ValidationError>>;
+
+/// A handle to the async process serving the validation host requests.
+pub struct ValidationHost {
+	to_host_tx: Mutex<mpsc::Sender<ToHost>>,
```

Comment:
Why do we need the mutex?

Reply:
Just to not require `&mut` on methods of `ValidationHost`.

Reply:
But why? xD

Reply:
Changed it to be &mut and no Mutex.